src/query/query_visitor.rs

```diff
@@ -0,0 +1,227 @@
+//! Query syntax tree traversal.
+//!
+//! Each method of [`QueryVisitor`] is a hook that can be overridden to customize the behavior when
+//! visiting the corresponding type of node. By default, the methods don't do anything. The actual
+//! walking of the ast is done by the `walk_*` functions. So to run a visitor over the whole
+//! document you should use [`walk_document`].
+//!
+//! Example:
+//!
+//! ```
+//! use graphql_parser::query::{
+//!     Field,
+//!     parse_query,
+//!     query_visitor::{QueryVisitor, walk_document},
+//! };
+//!
+//! struct FieldsCounter {
+//!     count: usize,
+//! }
+//!
+//! impl FieldsCounter {
+//!     fn new() -> Self {
+//!         Self { count: 0 }
+//!     }
+//! }
+//!
+//! impl<'ast> QueryVisitor<'ast> for FieldsCounter {
+//!     fn visit_field(&mut self, node: &'ast Field) {
+//!         self.count += 1
+//!     }
+//! }
+//!
+//! fn main() {
+//!     let mut number_of_type = FieldsCounter::new();
+//!
+//!     let doc = parse_query(r#"
+//!         query TestQuery {
+//!             users {
+//!                 id
+//!                 country {
+//!                     id
+//!                 }
+//!             }
+//!         }
+//!     "#).expect("Failed to parse query");
+//!
+//!     walk_document(&mut number_of_type, &doc);
+//!
+//!     assert_eq!(number_of_type.count, 2);
```

Comment:
This doesn't look right. There are four fields if visiting recursively: `query, id, country, id` and single one if not (`users`). Currently it looks like you just counting `users` twice.